File: Sorting Algorithms/heap_sort.cpp
/* Heap sort is a comparison based sorting technique based on Binary Heap data structure. It is similar to selection sort where we first find the maximum element 
and place the maximum element at the end. We repeat the same process for the remaining elements.
*/

// C++ program for implementation of Heap Sort

#include <iostream>
 
using namespace std;
 
// A function to heapify the array.
void MaxHeapify(int a[], int i, int n)
{
	int j, temp;
	temp = a[i];
	j = 2*i;
 
 	while (j <= n)
	{
		if (j < n && a[j+1] > a[j])
		j = j+1;
		// Break if parent value is already greater than child value.
		if (temp > a[j])
			break;
		// Switching value with the parent node if temp < a[j].
		else if (temp <= a[j])
		{
			a[j/2] = a[j];
			j = 2*j;
		}
	}
	a[j/2] = temp;
	return;
}
void HeapSort(int a[], int n)
{
	int i, temp;
	for (i = n; i >= 2; i--)
	{
		// Storing maximum value at the end.
		temp = a[i];
		a[i] = a[1];
		a[1] = temp;
		// Building max heap of remaining element.
		MaxHeapify(a, 1, i - 1);
	}
}
void Build_MaxHeap(int a[], int n)
{
	int i;
	for(i = n/2; i >= 1; i--)
		MaxHeapify(a, i, n);
}
int main()
{
	int n, i;
	cout<<"\nEnter the number of data element to be sorted: ";
	cin>>n;
	n++;
	int arr[n];
	for(i = 1; i < n; i++)
	{
		cout<<"Enter element "<<i<<": ";
		cin>>arr[i];
	}
	// Building max heap.
	Build_MaxHeap(arr, n-1);
	HeapSort(arr, n-1);
 
	// Printing the sorted data.
	cout<<"\nSorted Data ";
 
	for (i = 1; i < n; i++)
		cout<<"->"<<arr[i];
 
	return 0;
}
/* Output
Enter the number of data element to be sorted: 10
Enter element 1: 9
Enter element 2: 6
Enter element 3: 4
Enter element 4: 3
Enter element 5: 8
Enter element 6: 7
Enter element 7: 5
Enter element 8: 2
Enter element 9: 0
Enter element 10: 1
 
Sorted Data ->0->1->2->3->4->5->6->7->8->9
/*

File: Sorting Algorithms/heap_sort.py
'''
HEAP SORT
It is a sorting technique based on the heap data structure.
Heap is a complete binary tree, in which every level except the last, is filled
completely with its nodes, and nodes are far left.
We implement this sorting in an array using Max Heap, in which the parent node 
value is greater than it child node values. 
'''
def heapify(arr, length, index):
    # base case
    # we will call this function until the largest number is the index...
    largest_num_index = index
    left_index = (index * 2) + 1
    right_index = (index * 2) + 2
 
    if(left_index < length and arr[index] < arr[left_index]):
        largest_num_index = left_index
 
    if(right_index < length and arr[largest_num_index] < arr[right_index]):
        largest_num_index = right_index
 
    # if index is not the largest, make it the largest!
    # and run it again!
    if(largest_num_index != index):
        arr[index], arr[largest_num_index] = arr[largest_num_index], arr[index]
        heapify(arr, length, largest_num_index)
 
 
def heap_sort(arr):
    # need array length to create indices
    length = len(arr)
 
    for index in range(length, -1, -1):
        # ask about functions modifying arrays without return value
        # build max heap
        heapify(arr, length, index)
 
    # for each sorted heap, swap the root and the last number
    for index in range(length - 1, 0, -1):
        arr[index], arr[0] = arr[0], arr[index]
        # then call heapify again with the new array
        heapify(arr, index, 0)
 
 
# Taking Elements to be Sorted

data = []
n = int(input("Enter elements to be sorted :"))
 
for i in range(n):
    data.append(int(input()))
    
# Sending element to get sorted
heap_sort(data)
 
# Priting Elements after Getting Sorted
for d in data:
    print(d, end = " ")


'''
INPUT
Enter elements to be sorted : 5
30
50
10
20
40

OUTPUT
10 20 30 40 50

Time Complexity:
Best Case: O (nlogn)
Average Case: O (nlogn)
Worst Case: O (nlogn)
Space Complexity: O (1)
'''

